### DISCUSSION TITLE
Trouble deploying with Remix

### DISCUSSION THREAD

**Discussion by QuintonC** [vercel/community#2325] [Help]
### Summary

I've been running into issues with my deployments on Remix where it states I'm missing node_modules for various entrypoints and other related files.

What's more strange is that the build will publish and the page is able to be viewed (there are some broken links as some pages are not built yet, but clicking on any of the blog posts or components pages will result in an error.

## Deployment

https://vercel.com/quintonc/website-remix/4NZsNTqG6xXBnwPXXvcwCYKopSV1

## Hosted deployment

https://website-remix-8vygbftiu-quintonc.vercel.app/

## Output

```log
11:14:35.279 | The path "@remix-run/react" is imported in app/entry.client.tsx but "@remix-run/react" was not found in your node_modules. Did you forget to install it?
11:14:35.316 | The path "react" is imported in app/entry.client.tsx but "react" was not found in your node_modules. Did you forget to install it?
11:14:35.319 | The path "react-dom/client" is imported in app/entry.client.tsx but "react-dom/client" was not found in your node_modules. Did you forget to install it?
11:14:35.320 | The path "react/jsx-runtime" is imported in app/entry.client.tsx but "react/jsx-runtime" was not found in your node_modules. Did you forget to install it?
11:14:35.335 | mdx-bundler is possibly an ESM only package and should be bundled with "serverDependenciesToBundle" in remix.config.js.
11:14:35.339 | The path "@remix-run/node" is imported in app/routes/posts/$.tsx but "@remix-run/node" was not found in your node_modules. Did you forget to install it?
11:14:35.349 | The path "mdx-bundler/client" is imported in app/routes/posts/$.tsx but "mdx-bundler/client" was not found in your node_modules. Did you forget to install it?
11:14:36.001 | The path "front-matter" is imported in app/utils/mdxLoader.tsx but "front-matter" was not found in your node_modules. Did you forget to install it?
11:14:36.097 | The path "zustand" is imported in app/state/useStore.ts but "zustand" was not found in your node_modules. Did you forget to install it?
11:14:36.112 | The path "rehype-pretty-code" is imported in app/utils/mdxLoader.tsx but "rehype-pretty-code" was not found in your node_modules. Did you forget to install it?
11:14:36.112 | The path "zustand/middleware" is imported in app/state/useStore.ts but "zustand/middleware" was not found in your node_modules. Did you forget to install it?
11:14:36.138 | The path "rehype-slug" is imported in app/utils/mdxLoader.tsx but "rehype-slug" was not found in your node_modules. Did you forget to install it?
11:14:36.139 | The path "@react-spring/web" is imported in app/components/TOC.tsx but "@react-spring/web" was not found in your node_modules. Did you forget to install it?
11:14:36.148 | The path "remark-gfm" is imported in app/utils/mdxLoader.tsx but "remark-gfm" was not found in your node_modules. Did you forget to install it?
11:14:36.149 | The path "react-dom" is imported in app/sections/Toaster.tsx but "react-dom" was not found in your node_modules. Did you forget to install it?
11:14:36.162 | The path "@mdx-js/mdx" is imported in app/utils/mdxLoader.tsx but "@mdx-js/mdx" was not found in your node_modules. Did you forget to install it?
11:14:36.188 | The path "uniqid" is imported in app/state/toasterSlice.ts but "uniqid" was not found in your node_modules. Did you forget to install it?
```

## package.json

```json
{
  "private": true,
  "sideEffects": false,
  "scripts": {
    "build": "remix build",
    "dev": "remix dev",
    "format": "prettier --write \"**/*.{ts,tsx,md}\"",
    "lint": "eslint '**/*.ts*'",
    "postinstall": "patch-package",
    "typecheck": "tsc"
  },
  "dependencies": {
    "@mdx-js/mdx": "^2.3.0",
    "@mdx-js/react": "^2.3.0",
    "@react-spring/web": "^9.7.1",
    "@remix-run/node": "^1.16.0",
    "@remix-run/react": "^1.16.0",
    "@remix-run/vercel": "^1.16.0",
    "@vercel/node": "^2.6.2",
    "esbuild": "^0.17.16",
    "front-matter": "^4.0.2",
    "isbot": "latest",
    "mdx-bundler": "^9.2.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "rehype": "^12.0.1",
    "rehype-pretty-code": "^0.9.4",
    "rehype-slug": "^5.1.0",
    "remark-gfm": "^3.0.1",
    "shiki": "^0.14.2",
    "tailwindcss": "^3.3.1",
    "unified": "^10.1.2",
    "uniqid": "^5.4.0",
    "zustand": "^4.3.6"
  },
  "devDependencies": {
    "@esbuild-plugins/node-modules-polyfill": "^0.2.2",
    "@remix-run/dev": "^1.16.0",
    "@remix-run/eslint-config": "^1.16.0",
    "@remix-run/serve": "^1.16.0",
    "@shopify/prettier-config": "^1.1.2",
    "@types/react": "^18.0.25",
    "@types/react-dom": "^18.0.8",
    "@types/uniqid": "^5.3.2",
    "autoprefixer": "^10.4.14",
    "eslint": "^8.27.0",
    "patch-package": "^6.5.1",
    "postcss": "^8.0.9",
    "prettier": "^2.8.4",
    "prettier-plugin-tailwindcss": "^0.2.4",
    "typescript": "^4.8.4"
  },
  "engines": {
    "node": ">=14"
  }
}
```

### Remix config

```js
/** @type {import('@remix-run/dev').AppConfig} */
export default {
  ignoredRouteFiles: ['**/.*'],
  serverBuildPath: 'build/index.js',
  serverDependenciesToBundle: ['@mdx-js/react', 'mdx-bundler'],
  tailwind: true,
};
```

**Comment by QuintonC** [comment]
Update -- Specifying the environment variable for a vercel version seems to have resolve the issue in the build logs.

In order to get the deployments to build properly, I have to use `vercel@28.16.6`

Is this potentially a regression that was introduced **after** https://github.com/vercel/vercel/pull/9620 shipped?

**Reply by roniaxe** [reply]
> Update -- Specifying the environment variable for a vercel version seems to have resolve the issue in the build logs.
> 
> In order to get the deployments to build properly, I have to use `vercel@28.16.6`
> 
> Is this potentially a regression that was introduced **after** [vercel/vercel#9620](https://github.com/vercel/vercel/pull/9620) shipped?

I'm getting those build warnings on Vercel build, with a remix app, can you elaborate a bit on how you resolved it? how do you specify the environment variable for a vercel version ?

**Reply by QuintonC** [reply]
> I'm getting those build warnings on Vercel build, with a remix app, can you elaborate a bit on how you resolved it? how do you specify the environment variable for a vercel version ?

@roniaxe I resolved it by setting up environment variables and specifically setting the version number. To do this, follow these steps:

1. Click 'Settings' when in your Vercel project.
2. Click 'Environment Variables'
3. Enter `VERCEL_CLI_VERSION` for the `Key`
4. Enter `vercel@28.16.6` for the `Value`

However, I'm unsure if this is recommended or not. Builds take longer as a result because each deployment pulls the Vercel CLI version specified and then installs it. This took my builds from being ~20 seconds to being closer to ~1 minute on average. 

<img width="854" alt="image" src="https://github.com/vercel/community/assets/4250423/ba516c90-0b69-49af-9882-af3375cb0533" />

This seems a bit strange to me that the specified version isn't cached from deployment to deployment. Is this expected behavior @leerob?

**Reply by roniaxe** [reply]
Thanks so much @QuintonC !
As you mentioned, these are build warnings, and the app seems to work, so I might be waiting for a Vercel fix instead of changing the version for now, although it's really annoying seeing those warning during the build.

**Comment by swarnava** [comment]
Hi,

Do you have a reporducible repository which we can look at?

**Reply by QuintonC** [reply]
> Hi,
> 
> Do you have a reporducible repository which we can look at?

Sure, let me make the repository open-source for the time being.

https://github.com/QuintonC/website-remix

**Comment by QuintonC** [comment]
I'm now running into a different issue with node_module resolution that I've not been able to track down.

I've tried a couple of things here.

## Change attempt #1

- [Deployment URL](https://vercel.com/quintonc/website-remix/5Am6a1AXbgZU89LBmWANYJhrTsBo)
- [Preview URL](https://website-remix-dr4pc8d4d-quintonc.vercel.app/)

In remix.config.js, I tried including the following for my `serverDependenciesToBundle`: 
```js
  serverDependenciesToBundle: [
    /^@?\@mdx-js\/mdx.*/,
    /^@?mdx-bundler.*/,
    /^@?shiki.*/,
    /.*/,
```

And this yielded the following error when attempting to navigate to a [post](https://website-remix-dr4pc8d4d-quintonc.vercel.app/posts/building-an-mdx-blog-with-remix) or [component](https://website-remix-dr4pc8d4d-quintonc.vercel.app/components/anchor) route:

```log
Error: The esbuild JavaScript API cannot be bundled. Please mark the "esbuild" package as external so it's not included in the bundle.

More information: The file containing the code for esbuild's JavaScript API (/var/task/build/index.js) does not appear to be inside the esbuild package on the file system, which usually means that the esbuild package was bundled into another file. This is problematic because the API needs to run a binary executable inside the esbuild package which is located using a relative path from the API code to the executable. If the esbuild package is bundled, the relative path will be incorrect and the executable won't be found.
    at esbuildCommandAndArgs (/var/task/build/index.js:31173:15)
    at ensureServiceIsRunning (/var/task/build/index.js:31301:29)
    at Object.build (/var/task/build/index.js:31225:52)
    at bundleMDX2 (/var/task/build/index.js:57019:50)
    at getMDX (/var/task/build/index.js:73941:16)
    at loader4 (/var/task/build/index.js:74221:14)
    at Object.callRouteLoaderRR (/var/task/node_modules/@remix-run/server-runtime/dist/data.js:41:16)
    at callLoaderOrAction (/var/task/node_modules/@remix-run/router/dist/router.cjs.js:3817:16)
    at async Promise.all (index 0)
    at loadRouteData (/var/task/node_modules/@remix-run/router/dist/router.cjs.js:3379:19)
Caught exception while attempting to invoke processor for path: ---
meta:
  date: 2023-05-01
  description: I started rebuilding my blog in Remix and wanted to utilize MDX for an easier blogging format. This post details my experience in doing so, and some helpful information for others who might want to do the same.
  tag: Development
  slug: /posts/building-an-mdx-blog-with-remix
  title: Building an MDX blog with Remix

headers:
  Cache-Control: no-cache
---

## Introduction

I started rebuilding my blog in Remix and wanted to utilize MDX for an easier blogging format. This post details my experience in doing so, and some helpful information for others who might want to do the same.

Error: The esbuild JavaScript API cannot be bundled. Please mark the "esbuild" package as external so it's not included in the bundle.

More information: The file containing the code for esbuild's JavaScript API (/var/task/build/index.js) does not appear to be inside the esbuild package on the file system, which usually means that the esbuild package was bundled into another file. This is problematic because the API needs to run a binary executable inside the esbuild package which is located using a relative path from the API code to the executable. If the esbuild package is bundled, the relative path will be incorrect and the executable won't be found.
    at esbuildCommandAndArgs (/var/task/build/index.js:31173:15)
    at ensureServiceIsRunning (/var/task/build/index.js:31301:29)
    at Object.build (/var/task/build/index.js:31225:52)
    at bundleMDX2 (/var/task/build/index.js:57019:50)
    at getMDX (/var/task/build/index.js:73941:16)
    at loader4 (/var/task/build/index.js:74221:14)
    at Object.callRouteLoaderRR (/var/task/node_modules/@remix-run/server-runtime/dist/data.js:41:16)
    at callLoaderOrAction (/var/task/node_modules/@remix-run/router/dist/router.cjs.js:3817:16)
    at async Promise.all (index 0)
    at loadRouteData (/var/task/node_modules/@remix-run/router/dist/router.cjs.js:3379:19)
```

## Change attempt #2 

- [Deployment URL](https://vercel.com/quintonc/website-remix/Fhhu1wrE6h6ac1SaeUWSM6KLHURd)
- [Preview URL](https://website-remix-kd7qavwyq-quintonc.vercel.app/)

Trim down the `serverDependenciesToBundle` and see if we can get the information to load without specifying any dependencies (aside from `mdx-bundler`.

Changes to `remix.config.js`:
```js
  serverDependenciesToBundle: [/^@?mdx-bundler.*/],
```

Error when navigating to a [component](https://website-remix-kd7qavwyq-quintonc.vercel.app/components/anchor) or [post](https://website-remix-kd7qavwyq-quintonc.vercel.app/posts/building-an-mdx-blog-with-remix):

```log
✘ [ERROR] Error: ENOENT: no such file or directory, open '/var/task/node_modules/shiki/themes/material-theme-darker.json' [plugin @mdx-js/esbuild]
_mdx_bundler_entry_point-81e36130-aecf-4a1f-aec4-8f901502e012.mdx:0:0:
0 │ ---
╵ ^
Unhandled Promise Rejection 	{"errorType":"Runtime.UnhandledPromiseRejection","errorMessage":"Error: ENOENT: no such file or directory, open '/var/task/node_modules/shiki/themes/material-theme-lighter.json'","reason":{"errorType":"Error","errorMessage":"ENOENT: no such file or directory, open '/var/task/node_modules/shiki/themes/material-theme-lighter.json'","code":"ENOENT","errno":-2,"syscall":"open","path":"/var/task/node_modules/shiki/themes/material-theme-lighter.json","stack":["Error: ENOENT: no such file or directory, open '/var/task/node_modules/shiki/themes/material-theme-lighter.json'"]},"promise":{},"stack":["Runtime.UnhandledPromiseRejection: Error: ENOENT: no such file or directory, open '/var/task/node_modules/shiki/themes/material-theme-lighter.json'","    at process.<anonymous> (file:///var/runtime/index.mjs:1189:17)","    at process.emit (node:events:525:35)","    at process.emit (/var/task/node_modules/source-map-support/source-map-support.js:516:21)","    at emit (node:internal/process/promises:149:20)","    at processPromiseRejections (node:internal/process/promises:283:27)","    at processTicksAndRejections (node:internal/process/task_queues:96:32)"]}
Unknown application error occurred
Runtime.Unknown
```

---

I also reached out in the Remix discord group to see if there was any additional input from folks, and I got a reply from Kent Dodds who [recommended trying this approach to debug](https://www.epicweb.dev/tips/use-isolation-to-solve-complex-problems) the `mdx-bundler` configuration to see if it was an issue with my options or configuration. However, when testing this, I was not able to replicate the error in my dev environment. 

I'm beginning to wonder if I'm missing an environment variable somewhere or if there is an issue with the way that the node_modules are resolved within a Vercel environment.

**Comment by swarnava** [comment]
We are investigating this currently. cc @TooTallNate

**Reply by QuintonC** [reply]
Amazing. Thanks!

**Reply by QuintonC** [reply]
Bumping this to see if there are any updates from the Vercel side 🤞 

@swarnava @TooTallNate

**Reply by swarnava** [reply]
We are preparing for a fix here: https://github.com/vercel/remix/commit/0508f9763df775861dec87deaaaef79252ad4fbd. We will update once this is landed. Thank you

**Reply by QuintonC** [reply]
> We are preparing for a fix here: [vercel/remix@0508f97](https://github.com/vercel/remix/commit/0508f9763df775861dec87deaaaef79252ad4fbd). We will update once this is landed. Thank you

Thanks! If you have a patch version you would like for me to test out let me know and I'll give it a try 👍

**Reply by swarnava** [reply]
You can set following environment variable on your project and give it a try:

`VERCEL_CLI_VERSION=vercel@29.2.1`

**Reply by QuintonC** [reply]
> You can set following environment variable on your project and give it a try:
> 
> `VERCEL_CLI_VERSION=vercel@29.2.1`

@swarnava  It looks like it's still erroring out when trying to load a page that is using `mdx-bundler` and `shiki`.

Both packages are defined in my `package.json`, but navigating to a page that runs a server function which compiles MDX content, I receive an error:

```log
✘ [ERROR] Error: ENOENT: no such file or directory, open '/var/task/node_modules/shiki/themes/material-theme-darker.json' [plugin @mdx-js/esbuild]
_mdx_bundler_entry_point-90701b04-9493-4bc6-8207-2464766fa489.mdx:0:0:
0 │ ---
╵ ^
Unhandled Promise Rejection 	{"errorType":"Runtime.UnhandledPromiseRejection","errorMessage":"Error: ENOENT: no such file or directory, open '/var/task/node_modules/shiki/themes/material-theme-lighter.json'","reason":{"errorType":"Error","errorMessage":"ENOENT: no such file or directory, open '/var/task/node_modules/shiki/themes/material-theme-lighter.json'","code":"ENOENT","errno":-2,"syscall":"open","path":"/var/task/node_modules/shiki/themes/material-theme-lighter.json","stack":["Error: ENOENT: no such file or directory, open '/var/task/node_modules/shiki/themes/material-theme-lighter.json'"]},"promise":{},"stack":["Runtime.UnhandledPromiseRejection: Error: ENOENT: no such file or directory, open '/var/task/node_modules/shiki/themes/material-theme-lighter.json'","    at process.<anonymous> (file:///var/runtime/index.mjs:1189:17)","    at process.emit (node:events:525:35)","    at process.emit (/var/task/node_modules/source-map-support/source-map-support.js:516:21)","    at emit (node:internal/process/promises:149:20)","    at processPromiseRejections (node:internal/process/promises:283:27)","    at processTicksAndRejections (node:internal/process/task_queues:96:32)"]}
Unknown application error occurred
Runtime.Unknown
```

Deployment URL: https://vercel.com/quintonc/website-remix/EeuiBi84mnd4xS2FVhtXEjneSKSM

**Reply by swarnava** [reply]
Hi,

Do you have a updated repo which we can look into?

**Reply by QuintonC** [reply]
> Hi,
> 
> Do you have a updated repo which we can look into?

@swarnava The repo which is in question is public -- https://github.com/QuintonC/website-remix

**Reply by swarnava** [reply]
Hi,

Could you try :

>VERCEL_CLI_VERSION=vercel@29.3.0

**Reply by QuintonC** [reply]
> Hi,
> 
> 
> 
> Could you try :
> 
> 
> 
> >VERCEL_CLI_VERSION=vercel@29.3.0

@swarnava still getting the same error when navigating to an MDX-based page (such as [this one](https://website-remix-j6hz496xc-quintonc.vercel.app/posts/building-an-mdx-blog-with-remix)). 

Deployment: https://vercel.com/quintonc/website-remix/45E7wBLxxLzRPEbr5PKxW4SL9Het

Error message is the same as provided earlier. `shiki` is listed as a dependency, so the module should exist. It seems something might be wrong with the location of `node_modules`. 

```

✘ [ERROR] Error: ENOENT: no such file or directory, open '/var/task/node_modules/shiki/themes/material-theme-darker.json' [plugin @mdx-js/esbuild]
_mdx_bundler_entry_point-f5d53ee1-cd74-4c5f-8a1a-4a33c59d7941.mdx:0:0:
0 │ ---
╵ ^
Unhandled Promise Rejection 	{"errorType":"Runtime.UnhandledPromiseRejection","errorMessage":"Error: ENOENT: no such file or directory, open '/var/task/node_modules/shiki/themes/material-theme-lighter.json'","reason":{"errorType":"Error","errorMessage":"ENOENT: no such file or directory, open '/var/task/node_modules/shiki/themes/material-theme-lighter.json'","code":"ENOENT","errno":-2,"syscall":"open","path":"/var/task/node_modules/shiki/themes/material-theme-lighter.json","stack":["Error: ENOENT: no such file or directory, open '/var/task/node_modules/shiki/themes/material-theme-lighter.json'"]},"promise":{},"stack":["Runtime.UnhandledPromiseRejection: Error: ENOENT: no such file or directory, open '/var/task/node_modules/shiki/themes/material-theme-lighter.json'","    at process.<anonymous> (file:///var/runtime/index.mjs:1189:17)","    at process.emit (node:events:525:35)","    at process.emit (/var/task/node_modules/source-map-support/source-map-support.js:516:21)","    at emit (node:internal/process/promises:149:20)","    at processPromiseRejections (node:internal/process/promises:283:27)","    at processTicksAndRejections (node:internal/process/task_queues:96:32)"]}
Unknown application error occurred
Runtime.Unknown
```

**Reply by QuintonC** [reply]
@swarnava Sending a ping just to make sure that you saw this last comment ^

**Reply by swarnava** [reply]
We have fixed the initial reported issue. can you create separate discussion for the newer one you are experiencing?thanks

**Reply by QuintonC** [reply]
> We have fixed the initial reported issue. can you create separate discussion for the newer one you are experiencing?thanks

Really? 😕

That's not very helpful given the issue is titled "Trouble deploying with Remix." Kind of a bummer to get a response like that after troubleshooting this endlessly for the past two weeks.

**Reply by TooTallNate** [reply]
@QuintonC Regarding the `shiki` dependency, it appears to be loading files in a way that is not compatible with our static file analysis. You _may_ be able to make it work by specifying the path to the module in `includeFiles` configuration, so that all the files within that directory are included within the serverless function.

**Reply by QuintonC** [reply]
> @QuintonC Regarding the `shiki` dependency, it appears to be loading files in a way that is not compatible with our static file analysis. You _may_ be able to make it work by specifying the path to the module in `includeFiles` configuration, so that all the files within that directory are included within the serverless function.

@TooTallNate I tried at one point to specify the `includeFiles` param inside of `vercel.json`, but that ultimately failed because of a different issue -- that being that there is an issue with detecting serverless functions for Remix builds with Vercel. 

I **did** find a temporary solution to this, but I don't think it's a good one, and I'd think that we can work together to find a better solution to the issue as I've seen many other folks running into a similar issue with Next.js **and** Remix.

That aside, here is some information for a deploy where I attempted to provide configuration for this -

Commit: https://github.com/QuintonC/website-remix/commit/a3f139f59d843b87c841313450c66f1ffd1582a4
Deployment: https://vercel.com/quintonc/website-remix/CnpX5ZyckBY53oGhPjcnEK4qNMp9
Output ⬇️ 

```log
Vercel CLI 29.3.0
--
23:46:18.722 | Error: The pattern "components/*" defined in `functions` doesn't match any Serverless Functions inside the `api` directory.
```

However, looking at the deployment just before the attempt to provide configuration([deployment](https://vercel.com/quintonc/website-remix/HxgxrSGqx7Uk2PM7mdmgH9wz3cnX)), we can see that `components/*` **is** a serverless function

<img width="1156" alt="image" src="https://github.com/vercel/community/assets/4250423/9183bc1a-1ad9-4cdf-b0d0-5934e79c6bd9" />

Nonetheless, I managed to find a "hack" around this by providing an explicit path directive where I've copied over shiki languages and themes to my project (less than ideal).

Specifying a path to where I've copied the files over and providing those paths to the shiki highlighter manually resolves the issue, but I still don't think this is the ideal solution. Ideally, I'd be able to create a `vercel.json` file and specify my serverless functions with an `includeFiles` key value.

Temporary solution:

```ts
const SHIKI_PATH = `${__dirname}/../app/mdx-plugins/shiki`;

const getHighlighter: RehypePrettyCodeOptions['getHighlighter'] = async (
  options,
) => {
  touchShikiPath();

  const highlighter = await shiki.getHighlighter({
    // This is technically not compatible with shiki's interface but
    // necessary for rehype-pretty-code to work
    // - https://rehype-pretty-code.netlify.app/ (see Custom Highlighter)
    ...(options as any),
    paths: {
      languages: `${SHIKI_PATH}/languages/`,
      themes: `${SHIKI_PATH}/themes/`,
    },
  });

  return highlighter;
};
```

Note that this creates a larger build as it copies all of the languages + themes into each function's output directory.
